### DISCUSSION TITLE
[10월 1주차] 올바른 괄호

### DISCUSSION THREAD

**Discussion by z1nun** [z1nun/ASTERISK-#26] [알고리즘 문제풀이]
<html>
<body>
<h6 class="guide-section-title" style="box-sizing: border-box; padding: 0px; margin: 0px 0px 1rem; color: white; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-variant-numeric: inherit; font-variant-east-asian: inherit; font-weight: 700; font-stretch: inherit; font-size: 14px; line-height: 1.5; font-family: Inter, NotoSansKR, &quot;Helvetica Neue&quot;, Helvetica, Arial, &quot;맑은 고딕&quot;, &quot;malgun gothic&quot;, 돋움, Dotum, sans-serif, &quot;Apple Color Emoji&quot;, &quot;Noto Color Emoji&quot;; -webkit-font-smoothing: antialiased; letter-spacing: normal; orphans: 2; text-align: left; text-indent: 0px; text-transform: none; white-space: normal; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; background-color: rgb(38, 55, 71); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial; user-select: auto;">문제 설명</h6><div class="markdown solarized-dark" style="box-sizing: border-box; padding: 0px; margin: 0px; color: rgb(178, 192, 204); font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-variant-numeric: inherit; font-variant-east-asian: inherit; font-weight: 400; font-stretch: inherit; font-size: 16px; line-height: inherit; font-family: Inter, NotoSansKR, &quot;Helvetica Neue&quot;, Helvetica, Arial, &quot;맑은 고딕&quot;, &quot;malgun gothic&quot;, 돋움, Dotum, sans-serif, &quot;Apple Color Emoji&quot;, &quot;Noto Color Emoji&quot;; background-color: rgb(38, 55, 71); word-break: normal; overflow-wrap: break-word; letter-spacing: normal; orphans: 2; text-align: left; text-indent: 0px; text-transform: none; white-space: normal; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial; user-select: auto;"><p style="box-sizing: border-box; padding: 0px; margin: 0px; color: inherit; font-style: inherit; font-variant: inherit; font-weight: 400; font-stretch: inherit; font-size: inherit; line-height: 1.8; font-family: Inter, NotoSansKR, &quot;Helvetica Neue&quot;, Helvetica, Arial, &quot;맑은 고딕&quot;, &quot;malgun gothic&quot;, 돋움, Dotum, sans-serif, &quot;Apple Color Emoji&quot;, &quot;Noto Color Emoji&quot;; letter-spacing: -0.1px; user-select: auto; overflow-wrap: break-word;">괄호가 바르게 짝지어졌다는 것은 '(' 문자로 열렸으면 반드시 짝지어서 ')' 문자로 닫혀야 한다는 뜻입니다. 예를 들어</p><ul style="box-sizing: border-box; padding: 0px 0px 0px 2.5rem; margin: 1rem 0px 0px; color: inherit; font: inherit; user-select: auto;"><li style="box-sizing: border-box; padding: 0px; margin: 0px; color: inherit; font-style: inherit; font-variant: inherit; font-weight: inherit; font-stretch: inherit; font-size: 16px; line-height: 1.6; font-family: inherit; letter-spacing: -0.009em; list-style: inherit; user-select: auto;">"()()" 또는 "(())()" 는 올바른 괄호입니다.</li><li style="box-sizing: border-box; padding: 0px; margin: 0.5rem 0px 0px; color: inherit; font-style: inherit; font-variant: inherit; font-weight: inherit; font-stretch: inherit; font-size: 16px; line-height: 1.6; font-family: inherit; letter-spacing: -0.009em; list-style: inherit; user-select: auto;">")()(" 또는 "(()(" 는 올바르지 않은 괄호입니다.</li></ul><p style="box-sizing: border-box; padding: 0px; margin: 0.75rem 0px 0px; color: inherit; font-style: inherit; font-variant: inherit; font-weight: 400; font-stretch: inherit; font-size: inherit; line-height: 1.8; font-family: Inter, NotoSansKR, &quot;Helvetica Neue&quot;, Helvetica, Arial, &quot;맑은 고딕&quot;, &quot;malgun gothic&quot;, 돋움, Dotum, sans-serif, &quot;Apple Color Emoji&quot;, &quot;Noto Color Emoji&quot;; letter-spacing: -0.1px; user-select: auto; overflow-wrap: break-word;">'(' 또는 ')' 로만 이루어진 문자열 s가 주어졌을 때, 문자열 s가 올바른 괄호이면 true를 return 하고, 올바르지 않은 괄호이면 false를 return 하는 solution 함수를 완성해 주세요.</p><h5 style="box-sizing: border-box; padding: 0px; margin: 1.5rem 0px 0px; color: white; font-style: inherit; font-variant: inherit; font-weight: 400; font-stretch: inherit; font-size: 15px; line-height: 1.5; font-family: Inter, NotoSansKR, &quot;Helvetica Neue&quot;, Helvetica, Arial, &quot;맑은 고딕&quot;, &quot;malgun gothic&quot;, 돋움, Dotum, sans-serif, &quot;Apple Color Emoji&quot;, &quot;Noto Color Emoji&quot;; letter-spacing: -0.009em; user-select: auto;">제한사항</h5><ul style="box-sizing: border-box; padding: 0px 0px 0px 2.5rem; margin: 0.75rem 0px 0px; color: inherit; font: inherit; user-select: auto;"><li style="box-sizing: border-box; padding: 0px; margin: 0px; color: inherit; font-style: inherit; font-variant: inherit; font-weight: inherit; font-stretch: inherit; font-size: 16px; line-height: 1.6; font-family: inherit; letter-spacing: -0.009em; list-style: inherit; user-select: auto;">문자열 s의 길이 : 100,000 이하의 자연수</li><li style="box-sizing: border-box; padding: 0px; margin: 0.5rem 0px 0px; color: inherit; font-style: inherit; font-variant: inherit; font-weight: inherit; font-stretch: inherit; font-size: 16px; line-height: 1.6; font-family: inherit; letter-spacing: -0.009em; list-style: inherit; user-select: auto;">문자열 s는 '(' 또는 ')' 로만 이루어져 있습니다.</li></ul><hr style="box-sizing: content-box; padding: 0px; margin: 1.5rem 0px; color: inherit; font: inherit; height: 0px; overflow: visible; border-width: 0.0625rem 0px 0px; border-right-style: initial; border-bottom-style: initial; border-left-style: initial; border-color: rgb(23, 35, 52); border-image: initial; border-top-style: solid; user-select: auto;"><h5 style="box-sizing: border-box; padding: 0px; margin: 1.5rem 0px 0px; color: white; font-style: inherit; font-variant: inherit; font-weight: 400; font-stretch: inherit; font-size: 15px; line-height: 1.5; font-family: Inter, NotoSansKR, &quot;Helvetica Neue&quot;, Helvetica, Arial, &quot;맑은 고딕&quot;, &quot;malgun gothic&quot;, 돋움, Dotum, sans-serif, &quot;Apple Color Emoji&quot;, &quot;Noto Color Emoji&quot;; letter-spacing: -0.009em; user-select: auto;">입출력 예</h5>

s | answer
-- | --
"()()" | true
"(())()" | true
")()(" | false
"(()(" | false

<h5 style="box-sizing: border-box; padding: 0px; margin: 1.5rem 0px 0px; color: white; font-style: inherit; font-variant: inherit; font-weight: 400; font-stretch: inherit; font-size: 15px; line-height: 1.5; font-family: Inter, NotoSansKR, &quot;Helvetica Neue&quot;, Helvetica, Arial, &quot;맑은 고딕&quot;, &quot;malgun gothic&quot;, 돋움, Dotum, sans-serif, &quot;Apple Color Emoji&quot;, &quot;Noto Color Emoji&quot;; letter-spacing: -0.009em; user-select: auto;">입출력 예 설명</h5><p style="box-sizing: border-box; padding: 0px; margin: 0.75rem 0px 0px; color: inherit; font-style: inherit; font-variant: inherit; font-weight: 400; font-stretch: inherit; font-size: inherit; line-height: 1.8; font-family: Inter, NotoSansKR, &quot;Helvetica Neue&quot;, Helvetica, Arial, &quot;맑은 고딕&quot;, &quot;malgun gothic&quot;, 돋움, Dotum, sans-serif, &quot;Apple Color Emoji&quot;, &quot;Noto Color Emoji&quot;; letter-spacing: -0.1px; user-select: auto; overflow-wrap: break-word;">입출력 예 #1,2,3,4<br style="box-sizing: border-box; padding: 0px; margin: 0px; color: inherit; font: inherit; user-select: auto;">문제의 예시와 같습니다.</p></div>
</body>
</html>

https://school.programmers.co.kr/learn/courses/30/lessons/12909

**Comment by FordangIT** [comment]
테스트는 통과하고, 제출에는 실패한 내 코드 

```
function solution(s){
// 닫히고 열었을 때 그 때 열고 닫힌 개수가 동일하면 괜찮아 
// 닫히고 열었을 때 열고 닫히 숫자의 개수가 다르면 안돼 
    let open =0;
    let close=0;
for(let i=0; i<s.length; i++){
    if(s[i]===')' && s[i+1]==='('){
    for(let j=0; j<=i; j++){
        if(s[j]==='('){
        open++;
    }else{
        close++;
    }
  }
  return open === close ? true : false; 
    }
}
}
```


다른 사람 풀이 참고해서 내가 작성한 코드 

```
function solution(s){
let count = 0; 
    // '(' 면  count+ 1 ')'일때 count-1
    // count < 0 일 때 false
    // 마지막에 count가 0 이어야 true : false
    for(let i of s){
        i==='(' ? count++ : count--;
        if(count<0) return false;
    } return count===0 ?  true : false
}
```

**Comment by 2Zerozero** [comment]
테스트 4 5 11 17 실패 / 효율성 테스트 2 실패

```
function solution(s){
    var answer = false;
    
    // [실행] 버튼을 누르면 출력 값을 볼 수 있습니다.
    console.log(s)
    // 괄호가 똑바로 닫히면 () ture
    // 괄호가 똑바도 안닫히면 ()) )() false
    
    // 일단, 시작과 끝이 () 로 끝나야한다.
    // ( 로 시작되면, 다음 자리에 ) 오는지 알아본다.
    for(let i = 1; i < s.length; i++) {
        if(s[0] === '(' && s[s.length-1]) {
            if(s[i] === ')' && s[i-1] === '(') {
                return answer = true;
            }   
        }
        if(s[s.length-1] === '(') {
            return answer = false;
        }
    }
    
    return answer;
}
```

**Comment by sweetyr928** [comment]
```js
function solution(s){
    if(s[0]===')')
        return false;
    let count = 0;
    for(let i=0;i<s.length;i++){
      if(count < 0) return false;

      s[i]==='(' ? count++ : count--;
    }
 
    return !count
}
```

**Comment by z1nun** [comment]
정확성 테스트 5,11,17 실패
효율성 테스트 2 실패 
```javascript
function solution(s){
    var answer = true;
    
//     // s는 문자열
//     // split('') 하여 배열로 변경한다.
//     // 첫번째 인덱스가 '(' && ')'이면 ture;
//     // 그 외 false
    
    const arr = s.split('')
    if (arr[0] === '(' && arr[arr.length-1] === ')'){
        return answer;
    } else {
        answer = false;
        return answer;
    }
}
```


# 카테고리의 stack/queue로 접근
LIFO, FIFO, push & pop! 스택과 큐를 이용해서 문제를 풀어보세요.

- https://velog.io/@longroadhome/%EC%9E%90%EB%A3%8C%EA%B5%AC%EC%A1%B0-JS%EB%A1%9C-%EA%B5%AC%ED%98%84%ED%95%98%EB%8A%94-.%ED%81%90-Queue

- https://blog.tomclansys.com/69

큐 가장 먼저 들어온 데이터가 가장 먼저 빠져나가야 하는 구조이며, 이는 스택과 상반되는 순서를 가진다.

### 첫번째 시도 (제대로 실행되지도 않음)
```javascript
function solution(s){

    // 올바른 괄호는 ( 로 열리면 반드시 ) 로 닫아야한다.
    // 반복문을 사용하여 앞에가 ( 이면 요소중에서 )를 삭제한다.
    // 결국에 빈배열이 되면 true
    // 빈 배열이 되지 않으면 false

    // 문자열을 split하여 배열로 만든다.
    // 배열 내에 요소를 정렬한다.
    // 반복문을 사용하여 요소를 조건에 맞게 삭제한다.
    // 빈 배열(배열의 길이가 0)이면 true;
    // 빈 배열이 아니면(배열의 길이가 0이 아니면) false;
    
    const arr = s.split('');
    arr.sort();

    let count = 0;

    while(count < 100000){
        if (arr.shift() === '(' && arr.pop() === ')' ) {
            if(arr.length === 0) {
                return true;
            } else if (arr.length === 1) {
                return false;
            }
        } 
        count++;
    }
}
```

### 두번째시도 (정확성, 효율성 통과)
```javascript
function solution(s){
    var answer = true;
    
    // (())()
    // ( 문자로 열렸으면 반드시 ) 문자로 닫혀야 한다.
    // 인덱스가 (면 빈배열에 (를 push 한다.
    // 인덱스가 )면 빈배열에 푸시한 (를 다시 pop한다.
    // 결국에 배열이 빈배열이 된다면 서로 쌍이 맞기 때문에 True
    // 그러나 배열이 빈배열이 아니라면 쌍이 맞지 않기 때문에 False
    // 만약 ) 로 시작한다면 ? ( 로 시작하는 예외사항을 flase


    const arr = [];
    
    for (let i of s) {
        if(i === '(') {
            arr.push(i)
        }else{ // )이면 
            arr.length > 0 ?   // 배열에 (가 있는지 먼저 확인
                arr.pop()
            : // 값이 양수가 아니면 ( 가 없는 비정상적인 배열
                answer = false;
            }
        }
    if (arr.length > 0) {
        answer = false;
    }
    return answer;
}
```

**Comment by iltae** [comment]
```js
function solution(s){
    // 문자열에 첫번째는 '('이고 마지막은 ')'이다.
    // '('와 ')'의 개수는 같아야 한다.
    let halfLen = s.split('').filter(el => el === '(').length;
    return s[0] === '(' && s[s.length-1] === ')' && halfLen * 2 === s.length;
}
```
테스트5와 테스트11이 실패..

**Comment by rhino-ty** [comment]
```js
// 안됨
// function solution(s) {
//   // 1. 0과 마지막은 각각 (와 )인 것만 걸러주기
//   // 2. 괄호 개수 같음
//   let result;
//   let count = 0;

//   s.split("").map((bracket) => (bracket === "(" ? count++ : count--));
//   count === 0 ? (result = true) : (result = false);

//   s[0] === "(" && s[s.length - 1] === ")" ? (result = true) : (result = false);
//   return result;
// }

function solution(s) {
  // 1. 0과 마지막은 각각 (와 )인 것만 걸러주기
  // 2. 괄호 개수 같음
  let count = 0;

  s.split("").map((bracket) => (bracket === "(" ? count++ : count--));

  return count === 0 && s[0] === "(" && s[s.length - 1] === ")" ? true : false;
}
```
// 실패!

**Comment by fairytale779** [comment]
```js
function solution(s){
    if (s[s.length -1] !== ')') return false;

    let answer = 0;
    for (let i = 0; i<s.length; i++) {
        s[i] === '(' ? answer++ : answer--;
        if (answer < 0) return false;
    }
    return (answer) ? false : true ;
}```
